### DISCUSSION TITLE
Webpack5: How to generate same moduleId between double webpack config file.

### DISCUSSION THREAD

**Discussion by childrentime** [webpack/webpack#17325] [Q&A]
## Problem

when I upgrade my webpack version to webpack5, I can not find a way to genarate the same moduleId between double webpack config. These config files are almost the same, but one for server side, one for client side, cause I need react ssr for my application. And some loaders use moduleId to remove duplicates.

```js
moduleIds: 'hashed' | 'deterministic' | 'named' | 'natural'
```

config above can not work.

**Comment by alexander-akait** [comment]
Please provide reproducible test repo and fill out the issue, some old loader and plugins can break a module ids

**Comment by childrentime** [comment]
https://github.com/childrentime/terser-react Thank you for checkout this🥰. Anyway I find if your rules between two config files is not consistent sequence, the moduleId of file will the diifferent. It took me a long time to fix it🥹.

**Comment by childrentime** [comment]
run `npm run start` and checkout your brower head tag you will reproduce it.

**Reply by alexander-akait** [reply]
@childrentime Just want to undestand, why you want to have the same IDs in your case for server and client?

**Comment by childrentime** [comment]
@alexander-akait It will cause duplicate css in this example. My original purpose was to solve this problem because I am upgrading projects in company business that I should be careful to anything to prevent mistakes. I used a different order of rules on the client side and the server side which caused this problem before, Now it's solved by use the same order.

**Reply by alexander-akait** [reply]
Sorry, you can't have the same module ids in your case (you have two independent builds, they don't guarantee that you will always have the same IDs between builds, even more plugins and loader can inject more modules and even change module IDs), and I want to say it never works for webpack v4 (if you have id optimizations, so it was a luck if it works without any problems), you can't rely on ID(s) at all, I can't provide a valid setup here, but `isomorphic-style-loader` should calculate module ID based on content in the loader, so for the same content you will have the same ID, technically you can calculate ID(s) on the client side too, but it is bad for perfomance, I really don't know why `isomorphic-style-loader` rely in such things.

**Comment by childrentime** [comment]
Thanks! `isomorphic-style-loader` has no maintainers at a period of time, maybe I will copy it and change the way to calc IDS. Thanks for your clear elaborate!

**Reply by alexander-akait** [reply]
@childrentime There is another solution - implement own ids plugin, you can copy paste the logic and use `static` property to store ids, so you can sync them between compilation, but be careful.

Here example https://github.com/webpack/webpack/blob/main/lib/ids/DeterministicModuleIdsPlugin.js, as you can see there is no a lot of code...

**Reply by childrentime** [reply]
thanks!

**Comment by childrentime** [comment]
@alexander-akait I am wondering, when writing a loader, if it's possible to generate a same unique ID for a specific module path without considering the webpack configuration

**Comment by childrentime** [comment]
```
    var css = require(${stringifyRequest(this, `!!${request}`)});
    var content = typeof css === 'string' ? [[module.id, css, '']] : css;
```

 In a pitch loader, there is a code snippet like this. Can I generate the same ID for content across different webpack configurations without relying on webpack's moduleId?

Maybe
```
const modulePath = this.resourcePath;
const moduleName = this.resourceQuery || '';
 const moduleId = `my-module-${modulePath}-${moduleName}`;
```

**Comment by childrentime** [comment]
I only want to solve this problem in one loader, is there a general way to do it
